crates/stackable-certs/src/ca/mod.rs

```diff
@@ -1,486 +1,47 @@
-                    x509_cert::builder::Error::Signature(_),
-                ) => panic!(
-                    "it is impossible to compare the opaque Error contained witin signature::error::Error"
-                ),
-                _ => false,
-            },
-            (lhs, rhs) => lhs == rhs,
-        }
-    }
-}
-
-/// Defines all error variants which can occur when loading a CA from a
-/// Kubernetes [`Secret`].
-#[derive(Debug, Snafu)]
-pub enum SecretError<E>
-where
-    E: std::error::Error + 'static,
-{
-    #[snafu(display("failed to retrieve secret \"{secret_ref}\""))]
-    GetSecret {
-        source: kube::Error,
-        secret_ref: SecretReference,
-    },
-
-    #[snafu(display("invalid secret type, expected {TLS_SECRET_TYPE}"))]
-    InvalidSecretType,
-
-    #[snafu(display("the secret {secret:?} does not contain any data"))]
-    NoSecretData { secret: ObjectRef<Secret> },
-
-    #[snafu(display("the secret {secret:?} does not contain TLS certificate data"))]
-    NoCertificateData { secret: ObjectRef<Secret> },
-
-    #[snafu(display("the secret {secret:?} does not contain TLS private key data"))]
-    NoPrivateKeyData { secret: ObjectRef<Secret> },
-
-    #[snafu(display("failed to read PEM-encoded certificate chain from secret {secret:?}"))]
-    ReadChain {
-        source: x509_cert::der::Error,
-        secret: ObjectRef<Secret>,
-    },
-
-    #[snafu(display("failed to parse UTF-8 encoded byte string"))]
-    DecodeUtf8String { source: std::str::Utf8Error },
-
-    #[snafu(display("failed to deserialize private key from PEM"))]
-    DeserializeKeyFromPem { source: E },
-}
+pub use k8s::*;
 
 /// A certificate authority (CA) which is used to generate and sign
-/// intermidiate or leaf certificates.
+/// intermediate or leaf certificates.
 #[derive(Debug)]
-pub struct CertificateAuthority<S>
+pub struct CertificateAuthority<SK>
 where
-    S: CertificateKeypair,
-    <S::SigningKey as signature::Keypair>::VerifyingKey: EncodePublicKey,
+    SK: CertificateKeypair,
+    <SK::SigningKey as signature::Keypair>::VerifyingKey: EncodePublicKey,
 {
-    certificate_pair: CertificatePair<S>,
+    certificate_pair: CertificatePair<SK>,
 }
 
 impl<S> CertificateAuthority<S>
 where
     S: CertificateKeypair,
     <S::SigningKey as signature::Keypair>::VerifyingKey: EncodePublicKey,
 {
-    /// Creates a new CA certificate with many parameters set to their default
-    /// values.
-    ///
-    /// These parameters include:
-    ///
-    /// - a randomly generated serial number
-    /// - a default validity of one hour (see [`DEFAULT_CA_VALIDITY_SECONDS`])
-    ///
-    /// The CA contains the public half of the provided `signing_key` and is
-    /// signed by the private half of said key.
-    ///
-    /// If the default values for the serial number and validity don't satisfy
-    /// the requirements of the caller, use [`CertificateAuthority::new_with`]
-    /// instead.
```

Comment:
note: Compared to these docs, the current docs lack in depth and technical explanations. I've spent a significant amount of time on these and would like to see them carried over to the new implementation as complete as possible.

Reply:
I tried carrying over everything that is still relevant: 1322779